packages/components/src/typography/typography.tsx

```diff
@@ -0,0 +1,124 @@
+import React, { ReactNode } from "react";
+import { styled } from "@linaria/react";
+
+type TypographyElement = "h1" | "h2" | "h3" | "h4" | "h5" | "h6" | "span" | "p";
+
+type TypographySize =
+  | "h1"
+  | "h2"
+  | "h3"
+  | "h4"
+  | "h5"
+  | "sm"
+  | "xs"
+  | "default";
+
+// Matches color token categories
+type TypographyColor =
+  | "alert"
+  | "success"
+  | "info"
+  | "warning"
+  | "brand"
+  | "black"
+  | "white";
+
+type TypographyProps = {
+  /**
+   * This prop can be used to specify which size text element should
+   * actually be rendered, in the case that you want to render an element
+   * as one heading but style it as if it were another. If both an `as` prop
+   * and a `size` prop are passed, the `as` will be used to determine the element
+   * and the `size` will be used to determine the classname.
+   */
+  as?: TypographyElement;
+  /**
+   * The text content to present.
+   */
+  children: ReactNode;
+  /**
+   * The color of the text element. If no color provided, a default color
+   * will be used.
+   */
+  color?: TypographyColor;
+  /**
+   * The size of the html element. If no `as` prop is provided, then
+   * the component uses this value to determine which html tag to render
+   * (e.g. 'h1', 'h2', etc.)
+   */
+  size: TypographySize;
+};
+
+const typography = styled.span`
+  color: ${({ color }) =>
+    color ? tokens.EdsColorAlert700 : tokens.EdsColorBlack};
+  font-family: ${tokens.EdsFontFamily};
+`;
+
+const H3 = styled(typography)`
+  font-size: ${tokens.EdsFontSizeH3};
+  line-height: ${tokens.EdsFontLineHeightH3};
+`;
+
+const Body = styled(typography)`
+  font-size: ${tokens.EdsFontSizeDefault};
+  line-height: ${tokens.EdsFontLineHeightDefault};
+`;
+
+function elementFromSize(size: TypographySize): TypographyElement {
+  switch (size) {
+    case "h1":
+    case "h2":
+    case "h3":
+    case "h4":
+    case "h5":
+      return size;
+    case "default":
+    case "sm":
+    case "xs":
+      return "p";
+    default:
+      throw new TypeError(`Received invalid size option '${size}'`);
+  }
+}
+
+function styledComponentFromSize(size: TypographySize): typeof typography {
+  switch (size) {
+    case "h3":
+      return H3;
+    case "default":
+      return Body;
+    default:
+      throw new TypeError(`Received invalid size option '${size}'`);
+  }
+}
+
+function Typography(props: TypographyProps): JSX.Element {
+  const {
+    as,
+    children,
+    color,
+    size,
+    /**
+     * Components that wrap typography sometimes requires props such as event handlers
+     * to be passed down into the element. One example is the tooltip component.  It
+     * attaches a onHover and onFocus event to the element to determine when to
+     * trigger the overlay.
+     */
+    ...rest
+  } = props;
+
+  const TypographyComponent = styledComponentFromSize(size);
+  return (
+    <TypographyComponent
+      as={as || elementFromSize(size)}
```

Comment:
pulled this from the Along component, but I can't tell how it works 🤔

Reply:
Are you referring to the `as` prop, specifically?

If so, `as` is a prop from styled-components, so I wouldn't expect it to work here.

The intent is that someone can specify that they want something to look like an h2 (or h1, h3, whatever), but actually be something else (to fit into the document structure).

Reply:
ohhh ok that makes much more sense. I was going crazy trying to figure out how `as` was actually being used